### PR TITLE
NuFast oscillation calculator

### PR DESCRIPTION
**Pull request by cullenmsullivan**
This PR incorporates an implementation of the [NuFast](https://arxiv.org/abs/2405.02400) algorithm by Peter Denton (FNAL) and Stephen Parke (BNL). The algorithm is a 3F unitarity-assuming no-NSI neutrino oscillation calculator optimized for long baseline experiments. The algorithm is similar in spirit to `OscCalcAnalytic`, except some approximations make finding eigenvectors of the Hamiltonian slightly faster.
Testing the algorithm, so far I have found that although Stan fits are not faster with `OscCalcNuFast` compared to `OscCalcAnalytic`, I have found that calculations for the `<double>` template are 20%~30% faster (see NOvA docdb-64478 and docdb-64815).
In this pull request, I have migrated the caching system of `OscCalcAnalytic` to a new file, `Cache.h` so that it can be shared between `OscCalcAnalytic` and `OscCalcNuFast`.

### CODE REVIEW

File: OscLib/OscCalcNuFast.cxx
Comment (chenel) [inline OscLib/OscCalcNuFast.cxx:136]: Might be worth a comment here that this works for both mass orderings because in IO `Dmsq32` and `Dmsq21` have opposite sign

File: OscLib/OscCalcNuFast.h
Comment (chenel) [inline OscLib/OscCalcNuFast.h]: Force-inlining these two quite-complex functions seems fairly dubious to me.  They are long enough---and `RecomputeProbabilityMatrix()` can even do an allocation in it (!)---that I would be surprised if inlining made any measurable difference.  Did you do any performance checking to verify that it helps?  Did you do any comparisons on the size of the output code to see if there was any impact?

File: OscLib/OscCalcAnalytic.h
Comment (chenel) [inline OscLib/OscCalcAnalytic.h]: just remove the line, rather than commenting out.  we can always resurrect later from Git if needed

File: OscLib/OscCalcAnalytic.h
Comment (chenel) [inline OscLib/OscCalcAnalytic.h]: similarly (just delete, don't comment)

File: OscLib/OscCalcAnalytic.h
Comment (chenel) [inline OscLib/OscCalcAnalytic.h]: and again (just delete, don't comment)

File: OscLib/OscCalcNuFast.cxx
Comment (chenel) [inline OscLib/OscCalcNuFast.cxx]: in C++11 an `~_OscCalcNuFast() = default;` in the class declaration is preferred to an empty explicit destructor implementation

File: OscLib/OscCalcNuFast.cxx
Comment (chenel) [inline OscLib/OscCalcNuFast.cxx:27]: `(void)` is a C-ism that's generally frowned upon in C++ for ideological reasons.  not a big deal, but looks weird
```suggestion
  _OscCalcNuFast<T>* _OscCalcNuFast<T>::Copy() const { return new _OscCalcNuFast<T>(*this); }
```

File: OscLib/OscCalcNuFast.cxx
Comment (chenel) [inline OscLib/OscCalcNuFast.cxx]: If this block is only used sometimes, it should be inside an `if` so that the derivatives aren't computed by Stan when not used

File: OscLib/OscCalcNuFast.h
Comment (chenel) [inline OscLib/OscCalcNuFast.h]: why is this guy a `double` and not an `unsigned int`, out of curiosity?

**Review by chenel — COMMENTED** [review]
In general, looks very nice, and is fine for a first implementation.  Most of my inline comments are simple ones.

Bigger question: I've only looked over `OscCalcNuFast` so far.  Is `OscCalcNuFastDumb` something you actually want to include in the final product?  It seems to be the same apart from the caching?

I have a couple of meta-thoughts that we should discuss, as well:

1. Rather than copy-pasting-editing the code from NuFast, it would be much neater if there were a `nufast::Calculator` object we could just make a member of the `OscLib` object and use it, thus avoiding duplicating the code from them.  Obviously the NuFast authors didn't structure their code this way, but I wonder if maybe they would be interested in doing such a thing if they knew we were using it.  And if so, I bet your templated version would not be a difficult sell.  I know Peter Denton and would be happy to initiate the conversation.
2. I'm a bit uneasy about the cache when applied to Stan types.  Stan uses an internal buffer for storing its autodiff stack that's not exposed to the user, and it can be cleared without warning (if the user does anything that involves actually computing the derivatives).  I've added guardrails to the usage with `StanFitter`, so that this is less likely to happen in practice, but we should maybe talk about how necessary the caching actually is.  (And all of this obviously applies to `OscCalcAnalytic`, which nobody has seriously used so far.)

File: OscLib/OscCalcNuFast.h
Comment (cullenmsullivan) [reply OscLib/OscCalcNuFast.h]: That's a good point. I didn't think very hard about aggressively inlining these functions because `OscCalcAnalytic` does similar things.
I didn't make any testing with and without inlining, but I highly doubt it's actually helpful in any measurable way. I want to remove the inlining and assume that it makes no change to runtime.

File: OscLib/OscCalcAnalytic.h
Comment (cullenmsullivan) [reply OscLib/OscCalcAnalytic.h]: Oops, this is an oversight from me. I forgot to tidy `OscCalcAnalytic` before making the PR - will fix.

File: OscLib/OscCalcNuFast.h
Comment (cullenmsullivan) [reply OscLib/OscCalcNuFast.h]: This is simply a mistake from me - good catch!

File: OscLib/OscCalcNuFast.cxx
Comment (cullenmsullivan) [reply OscLib/OscCalcNuFast.cxx]: I stuck this block inside an `if` and confirmed that the spectra still converge. Actually because I got curious, I just discovered that there is no (well, some extremely small) difference between 0 and 1 Newton iterations by testing spectra. I'm going to set the default to 0 Newton iterations.

File: OscLib/OscCalcNuFast.cxx
Comment (cullenmsullivan) [reply OscLib/OscCalcNuFast.cxx:27]: Actually, I prefer using `(void)`, and I wasn't aware of the modern C++ convention avoiding it. It makes more sense to me to write it with `(void)`, but this isn't a hill I care do die on or anything.

**Comment by cullenmsullivan** [comment]
I've addressed most of the inline changes. To address the larger issues:

1. Including `OscCalcNuFastDumb` in this PR is a mistake from me, and I never intended to actually commit it. It's strictly slower than `OscCalcNuFast` as-is, and I only used it to help benchmark my progress as I increasingly increased how aggressively things are cached. I've now removed it. With `OscCalcNuFastDumb` gone, there's no immediate need to refactor `OscCalcNuFast`, but it could improve the readability and extensibility (perhaps NSI/sterile folks would be interesting in adapting it?). When writing this, I tried to balance compatibility with `OscLib` but also preservation of the original algorithm as the authors wrote it. I'm open to collaboration with the authors to streamline the code more.
2. Avoiding use of the cache with Stan types would be an easy change, but from a very quick test, it seems that the speed of Stan MCMC sampling is sensitive to use of the cache. When running a fit side-by-side with and without a caching calculator, the caching fit ran about twice as quickly. This might need more study. Because `StanFitter` uses the `Eigen` templates for getting oscillation probabilities, the calculator shouldn't need to be queried for energy multiple times for one channel (right?). Perhaps the extra time is going into recomputing probabilities for multiple flavor -> flavor channels? If this is the case, we might be able to avoid the slowdown by saving all disappearance/appearance channels at the fitter level and also avoid use of a cache for Stan vars at the calculator level?

**Comment by chenel** [comment]
Thanks for studying the cache usefulness.  I think that given that it seems to matter, let's keep it for now and try to add guardrails later if we find people using the Stan-aware calculator irresponsibly. 😄   If you could add a warning in the comments, that would be great.

Let's write to Peter and get a feel for his level of interest in extending the code he has.  (The main advantage I can see of modularizing is if in the future he and his collaborators make some improvement, we may be able to adopt it without any tinkering on our end.)  I'm happy to start that conversation.

**Comment by chenel** [comment]
In the meantime, I think this is ready for regular review and incorporation into `OscLib`, so you can feel free to promote this out of 'draft' status.

**Comment by cullenmsullivan** [comment]
All right, I've marked it as ready for review. I think talking to Peter is a good idea as well - I'd appreciate it if you do start the conversation, thank you!

Following up from this PR (once merged), I'll make a sister PR in novasoft to save my testing macros and make an edit to `StanFitter` to incorporate the Stan-aware calculator.

**Review by chenel — DISMISSED** [review]
I'm happy with merging, per discussion above.

File: OscLib/OscCalcAnalytic.cxx
Comment (vhewes) [inline OscLib/OscCalcAnalytic.cxx]: since this function is now pulled in from `Cache.h`, this function should be deleted instead of just commented out

**Review by vhewes — CHANGES_REQUESTED** [review]
overall this looks good, i just have one change to request. i definitely agree with @chenel's comment about pulling this code in from NuFast if that's possible

File: OscLib/OscCalcAnalytic.cxx
Comment (cullenmsullivan) [reply OscLib/OscCalcAnalytic.cxx]: Oops, that's another piece I forgot to tidy - thanks for the good catch. I just committed a deletion

**Comment by gavinsdavies** [comment]
this just needs @vhewes to give approval to the requested changes.

**Review by vhewes — APPROVED** [review]
looks good! sorry for the delay

**Pull request merged**
